libbindgen/tests/tests.rs

```diff
@@ -0,0 +1,132 @@
+extern crate clap;
+extern crate diff;
+#[macro_use]
+extern crate env_logger;
+extern crate libbindgen;
+extern crate log;
+extern crate shlex;
+
+use std::env;
+use std::fs;
+use std::io::{BufRead, BufReader, Error, ErrorKind, Read};
+use std::path::{Path, PathBuf};
+
+#[path="../../src/options.rs"]
+mod options;
+use options::builder_from_flags;
+
+fn compare_generated_header(header: &PathBuf,
+                            builder: libbindgen::Builder)
+                            -> Result<(), Error> {
+    let file_name = try!(header.file_name()
+        .ok_or(Error::new(ErrorKind::Other, "spawn_bindgen expects a file")));
+
+    let mut expected = PathBuf::from(header);
+    expected.pop();
+    expected.pop();
+    expected.push("expectations");
+    expected.push("tests");
+    expected.push(file_name);
+    expected.set_extension("rs");
+
+    let output = match builder.generate() {
+        Ok(bindings) => bindings.to_string(),
+        Err(_) => "".to_string(),
+    };
+
+    let mut buffer = String::new();
+    let f = try!(fs::File::open(&expected));
+    let _ = try!(BufReader::new(f).read_to_string(&mut buffer));
+
+    if output == buffer {
+        return Ok(());
+    }
+
+    println!("diff expected generated\n--- expected: {:?}\n+++ generated from: {:?}",
+             expected,
+             header);
+    for diff in diff::lines(&buffer, &output) {
+        match diff {
+            diff::Result::Left(l) => println!("-{}", l),
+            diff::Result::Both(l, _) => println!(" {}", l),
+            diff::Result::Right(r) => println!("+{}", r),
+        }
+    }
+    Err(Error::new(ErrorKind::Other, "Header and binding differ!"))
+}
+
+fn create_bindgen_builder(header: &PathBuf)
+                          -> Result<libbindgen::Builder, Error> {
+    let source = try!(fs::File::open(header));
+    let reader = BufReader::new(source);
+
+    // Scoop up bindgen-flags from test header
+    let line: String = try!(reader.lines().take(1).collect());
+    let flags =
+        match line.split("bindgen-flags:").last().and_then(shlex::split) {
+            Some(args) => args,
+            _ => Vec::with_capacity(2),
+        };
+
+    // Fool builder_from_flags() into believing it has real env::args_os...
+    // - add "bindgen" as executable name 0th element
+    // - add header filename as 1st element
+    // - prepend raw lines so they're in the right order for expected output
+    // - append the test header's bindgen flags
+    let args = vec![
```

Comment:
Does this need to be a `vec!`. I think it should work just fine on the stack... Not too important though.